### PR TITLE
Linked Time: Add step selector to redux state Part 1

### PR DESCRIPTION
**Pull request by rileyajones**
## Motivation for features / changes
See #6028 for a full description of why this change is needed.
The amount of things needing to be changed for that to function correctly is too great for a single PR thus I am doing this in phases. This first phase is to create the redux logic needed.

## Technical description of changes
I updated the metrics reducer to set a min max and step selection value for each card when the tag data is loaded and we know their ids.

I added a new action and reducer to handle updates the card min max values.

I updated the time selection changed reducer to store the new time selection to redux. This works because the action is always called regardless of whether linked time is enabled.

I then updated the toggle range selection reducer to create an end step if needed (e.g. the user removed it).

## Screenshots of UI changes
N/A

## Detailed steps to verify changes work correctly (as executed by you)
All the tests should pass

## Alternate designs / implementations considered
I considered either using selectors with argument to get values for individual cards or closuring the cardId in a generated selector. I felt this solution of just fetching the map was the cleanest.

### CODE REVIEW

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (bmd3k) [inline tensorboard/webapp/metrics/store/metrics_reducers.ts:862]: My  understanding (which, admittedly, is weak and based on looking at metrics effects and data source code this morning) is that  a fetchTimeSeriesLoaded will only ever correspond to exactly one card.

It looks like here you are rebuilding the min/max for ALL cards on each response. Is that necessary? Can we limit it to only rebuilding min/max for the card represented in the response?

File: tensorboard/webapp/metrics/store/metrics_types.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_types.ts:162]: Yeah, I guess that makes more sense? All the types of cardIds are intercompatible because they are all just wrappers for string.  They are really only lexically different.
```
export type NonPinnedCardId = string;

export type PinnedCardId = string;

export type CardId = NonPinnedCardId | PinnedCardId;
```

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils_test.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils_test.ts:414]: I'm just going to add an `expect` to this test, I think the result should be about the same.

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils_test.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils_test.ts:727]: I've gone ahead and explicitly set the step data and broken it up into multiple runs.

File: tensorboard/webapp/metrics/store/metrics_types.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_types.ts:47]: These different data types already exist today but I agree this could use a cleanup. I would prefer to do this after this change has merged.
As we discussed there are ways to achieve this with `git`, however, GitHub does not have a good way of presenting PRs that are children of other open PRs. Due to this I'm going to hold off on this refactor until this series of PRs are merged.

File: tensorboard/webapp/metrics/store/metrics_types.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_types.ts:50]: I've gone ahead and moved this to `tensorboard/webapp/metrics/types.ts`. This actually allowed some additional BUILD file cleanups.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts]: I believe I've refactored this correctly AND made it more efficient at the same time.

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts
Comment (bmd3k) [inline tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts:388]: I think you can still get rid of the cast. Just:
` .map((stepDatum) => stepDatum.step); `

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (bmd3k) [inline tensorboard/webapp/metrics/store/metrics_reducers.ts:1074]: Could this instead be state.cardToMinMax[cardId].max?

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (bmd3k) [inline tensorboard/webapp/metrics/store/metrics_reducers.ts:1069]: Similarly, you need to set end: null for each time selection when you turn off range selection. You don't necessarily need to do that in this PR but keep it in mind.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (bmd3k) [inline tensorboard/webapp/metrics/store/metrics_reducers.ts:867]: Scalar cards should have runId null since they are "multi-run" cards. If you look at the state constructed for a scalar card with 4 runs, for example, you see that it gets 4 different entries in state.cardToTimeSelection. It should have a single entry.

See: https://cs.opensource.google/tensorflow/tensorboard/+/master:tensorboard/webapp/metrics/store/metrics_reducers.ts;l=120;drc=99e0e2788ee9f0bb95b3e08e5c018e394e3a18c8

If you need to construct individual cardMetadata objects, maybe its worth refactoring the logic out of buildCardMetadataList rather than trying to replicate it here.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (bmd3k) [inline tensorboard/webapp/metrics/store/metrics_reducers.ts:885]: I'm not certain but I don't think this is right.  What if user presses the reload button in the header (or even has auto-reload turned on) and kicks off another series of time series request/responses? Then will their previously-selected time selection be reset when the next response comes in?

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (bmd3k) [inline tensorboard/webapp/metrics/store/metrics_reducers.ts:862]: Thinking a little more about this, especially in light of the changes in https://github.com/tensorflow/tensorboard/pull/6175.

The isStepSelectorEnabled state is no longer a global indicator of whether we will show step selection for *all* cards or *zero* cards. You can have the checkbox turned "off" but still show step selection on a specific card. Likewise, you can have the checkbox turned "on" but still not show step selection on a specific card. This is something we have to keep in mind for this new ngrx-based time selection state, too, and I don't think we are taking that into account with these changes.

The way things are written now I suspect you planned to have timeSelection for all cards and just determine whether to show the steps in a card based on the isStepSelectorEnabled state.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (bmd3k) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts:1110]: That sounds great, thanks for clarifying.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (bmd3k) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts:873]: Your analysis makes sense. Not much harm in keeping it around.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts]: I still prefer to do it upfront. While the complexity is adding up it will at least all be upfront and in one place.

You are right, this did not correctly calculate the min/max for histograms and images cards, I didn't realize there each only contained a single run. 
I've updated this to only read the data from a single run for each of those card types.

Image cards are more complicated. I believe this is actually right for them despite the fact that the number of image cards is Tags * Runs * Samples the data only seems to be Tags * Runs.

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts:278]: I'm updating this pr to move `cardToTimeSelection` and `cardToMinMax` to be part of `cardStateMap`.

File: tensorboard/webapp/metrics/internal_types.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/internal_types.ts:35]: This was necessary before I stopped trying to support histogram and image cards.

File: tensorboard/webapp/metrics/store/metrics_selectors.ts
Comment (rileyajones) [inline tensorboard/webapp/metrics/store/metrics_selectors.ts:503]: This could technically just rely on the `linkedTimeSelection` being non-null because `getMetricsLinkedTimeSelection` returns `null` when linked time is not enabled. However, don't think that is obvious to a reader so I added this null check. I could imagine a comment having a similar effect.

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils_test.ts
Comment (rileyajones) [inline tensorboard/webapp/metrics/store/metrics_store_internal_utils_test.ts:443]: When a card is pinned, the new id is generated using `getPinnedCardId`. I'm opting to specify the returned value here, but it might read better of this line used the function instead
```suggestion
        [getPinnedCardId('card1')]: {
```

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (bmd3k) [inline tensorboard/webapp/metrics/store/metrics_reducers.ts:1104]: So here's some fun NgRx behavior that I've been exploring recently: NgRx will deeply call Object.freeze() on all state after your reducer is called.

What it means for you is that, at runtime, this L1104 will silently fail to modify `nextCardStateMap[cardId].timeSelection = linkedTimeSelection;`.

You won't be able to catch this in a test because the freeze is not applied in the tests.

I recommend changing L1101-1104 to:
```
nextCardStateMap[cardId] = {
  ...nextCardStateMap[cardId],
  timeSelection: linkedTimeSelection,
};
```

It has the benefit of actually modifying timeSelection AND it handles the case where nextCardStateMap[cardId] is undefined.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (bmd3k) [inline tensorboard/webapp/metrics/store/metrics_reducers.ts:875]: Using same reasoning as my other comment, L875-878 should be:

```
nextCardStateMap[cardId] = {
   ...nextCardStateMap[cardId],
  dataMinMax: nextMinMax,
};
```

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (bmd3k) [inline tensorboard/webapp/metrics/store/metrics_reducers.ts:1117]: This one might be ok but, I think, to be certain and to be consistent with other code, you could rewrite as:

```
nextCardStateMap[cardId] = {
 ...nextCardStateMap[cardId],
 userMinMax: minMax,
};
```

**Review by bmd3k — COMMENTED** [review]
This mostly looks good now but I think there are a few places where we need to adjust, accounting for NgRx's immutable state protections.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts:1104]: That is really interesting and makes a lot of sense. `Object.freeze` is an underutilized function.

The copy makes lots of sense and is more consistent, that is definitely how this should be done.

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts:333]: Good point, I think I went a little overboard on copying things.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts:1104]: Sure :+1:

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts:332]: I am worried about the pinned card state pointing to the same object as the original card. I'm not sure how the NgRx deep freeze you mentioned works so I'm hesitant to say it would resolve the issue.

I would rather developers not have to worry about whether newly added properties are copied over when a card is pinned.

I'm really not sure why I didn't think making a true deep copy was a good idea, the resulting code is much less ledgible and significantly more fragile with the only benefit being a slight performance boost. FWIW `structuredClone` is a slightly more performant solution than `JSON.parse(JSON.stringify())` but it's only supported by new browser versions so I'm intentionally not using it here.

File: tensorboard/webapp/metrics/store/metrics_reducers_test.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_reducers_test.ts:3057]: That was indeed hard to see, I'll move the tests.

File: tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts]: Here is a simple example of card entanglement via a shared state reference
1) The user expands the card.
`card1State = {tableExpanded: true}`
The card state map is now `{ card1: card1State}`
2) The user pins the card.
The card state map is now `{card1: card1State, pinnedCard1: card1State}`
3) The user collapses the pinned cards table
The card state map "unchanged"  `{card1: card1State, pinnedCard1: card1State}`
Except that `card1State = {tableExpanded: false}` thus collapsing the table for both cards.

Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts]: I don't mind the type assertion in this case. The object was initially `Partial<CardState>`, after being serialized and deserialized will still be `Partial<CardState>`

Comment (bmd3k) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts]: For the case you illustrate, I don't think NgRx will allow that to happen.

You can't do this, for example:
```
state.cardStateMap[pinnedCardId].tableExpanded = false
```

That will silently fail because of the frozen state. The tableExpanded property will continue to be true and neither card will collapse.

NgRx is basically forcing you to do something like this:

```
state.cardStateMap[pinnedCardId] = {
  ...state.cardStateMap[pinnedCardId],
  tableExpanded: false
};
```

That will work around the frozen state AND will decouple the pinned card's state from the original card's state. Pinned card will be collapsed but original card will be expanded.

Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_store_internal_utils.ts]: So to paraphrase, it is okay if both cards reference the same object temporarily because when an attribute changes the object will need to be reconstructed anyway. That makes sense, I guess I was overly worried about that intermediate state.

**Review by bmd3k — APPROVED** [review]

**Pull request merged**
